session: internal request execution API error types refactor

muzarski opened the pull request:
Ref: https://github.com/scylladb/scylla-rust-driver/issues/519

## run_request_once closure
To narrow the error return types of some important API functions (e.g. `RS::decide_should_retry` and `LBP::on_query_failure`), the error type of `execute_request_once` (previously known as `do_query`) needs to be adjusted as well. A couple of commits at the beginning of this PR do just that. What we achieve, is that we narrow the return error type of the closure from `QueryError` to `UserRequestError`.

## UserRequestError
This error is renamed to `RequestAttemptError` - as the name suggests, it represents a failure of a single attempt of execution.

## LBP and RetrySession
Both of these traits accept some error and do something based on the error. Before this PR, they would accept a `QueryError`, resulting in matches against variants that could never even be constructed before the call. In this PR, I pub-ified the `RequestAttemptError`, and made `LBP::on_query_failure` and `RetrySession::decide_should_retry` accept it instead of `QueryError`.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- ~[ ] I have adjusted the documentation in `./docs/source/`.~
- [x] I added appropriate `Fixes:` annotations to PR description.

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -1296,7 +1296,7 @@ impl Connection {
         &self,
         init_batch: &'b Batch,
         values: impl BatchValues,
-    ) -> Result<Cow<'b, Batch>, QueryError> {
+    ) -> Result<Cow<'b, Batch>, UserRequestError> {
         let mut to_prepare = HashSet::<&str>::new();
 
         {
```
Commit: "batch: narrow the return error type of prepare_batch "

🤔 It is a bit surprising that this change does not require any other changes. This is due to some conversions done by "?", right? I wonder if we should maybe limit such conversion.
If error A has a variant with type B, then conversion from B to A is imo fine. But this is not the case here, right? This is a conversion between two separate errors. Maybe it should be explicit?

muzarski replied:
`UserRequestError` has 11 variants. I think that `From<UserRequestError` for QueryError` impl comes in handy. I'm not a big fan of matching against all variants each time (assuming this is what you mean by explicit conversion).

Lorak-mmk replied:
That is not what I meant. I'm talking specifically about `From` trait, because this is what seems to be used by `?` operator and `into` calls. 
A method like `into_query_error` would be totally fine.

muzarski replied:
Okay, this makes sense. And I totally agree. Using `?` may lead to some unwanted implicit conversions.

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -1284,7 +1284,7 @@ impl Connection {
+                _ => Err(UserRequestError::UnexpectedResponse(
                     query_response.response.to_response_kind(),
                 )
                 .into()),
```
It took me a few seconds to notice the "into()" call, and now I'm more convinced that such conversions are a bad idea. Why not return the variant that we want explicitly instead of relying on conversions?

muzarski replied:
This `into()` call is later removed anyway (commit `batch: return UserRequestError from batch_with_consistency`). But I agree with the suggestion, and will start applying it in the future.

Lorak-mmk commented on scylla/src/transport/session.rs:
```diff
@@ -1650,7 +1650,7 @@ where
+        let run_query_result: RunRequestResult<NonErrorQueryResponse> = self
             .run_request(
                 statement_info,
                 &batch.config,
```
🔧 Previous PR (which is included in this one change the name from using "query " to "request" but here you are changing it back. It looks like a mistake.

muzarski replied:
Yeah, I wrote this code a while ago. The previous PR was created just today. I had to miss it during a rebase (or there had to be no conflict in this place, so I did not notice). Good catch

muzarski commented:
v1.1: Rebased on latest https://github.com/scylladb/scylla-rust-driver/pull/1156 and addressed @Lorak-mmk comments. Most important change is that we replaced `From<UserRequestError> for QueryError` with explicit conversion method `UserRequestError::into_query_error()`.

Lorak-mmk reviewed (approved):
Looks good.

A suggestion for the future (because this PR is probably not a good place for this): could you re-order variants in UserRequestError so that they match the order in QueryError? Changing the QueryError instead would be fine too - I just think that the variants present in both should appear in the same order in both.

wprzytula commented on scylla/src/transport/errors.rs:
```diff
@@ -908,6 +908,8 @@ pub(crate) enum UserRequestError {
     #[error("Reprepared statement's id does not exist in the batch.")]
     RepreparedIdMissingInBatch,
+    #[error("Failed to serialize query paramters: {0}")]
+    SerializationError(#[from] SerializationError),
```
:pick: typo: `paramters`

wprzytula commented on scylla/src/transport/errors.rs:
```diff
@@ -871,32 +871,52 @@ pub enum CqlEventHandlingError {
+/// on this error.
 #[derive(Error, Debug)]
-pub(crate) enum UserRequestError {
+#[non_exhaustive]
+pub enum UserRequestError {
+    /// Failed to serialize CQL request.
     #[error("Failed to serialize CQL request: {0}")]
     CqlRequestSerialization(#[from] CqlRequestSerializationError),
+
+    /// Database sent a response containing some error with a message
     #[error("Database returned an error: {0}, Error message: {1}")]
     DbError(DbError, String),
+
+    /// Received a RESULT server response, but failed to deserialize it.
     #[error(transparent)]
     CqlResultParseError(#[from] CqlResultParseError),
+
+    /// Received an ERROR server response, but failed to deserialize it.
     #[error("Failed to deserialize ERROR response: {0}")]
     CqlErrorParseError(#[from] CqlErrorParseError),
+
+    /// Received an unexpected response from the server.
     #[error(
         "Received unexpected response from the server: {0}. Expected RESULT or ERROR response."
     )]
     UnexpectedResponse(CqlResponseKind),
+
+    /// A connection has been broken during query execution.
     #[error(transparent)]
     BrokenConnectionError(#[from] BrokenConnectionError),
+
+    /// Failed to deserialize frame body extensions.
     #[error(transparent)]
     BodyExtensionsParseError(#[from] FrameBodyExtensionsParseError),
+
+    /// Driver was unable to allocate a stream id to execute a query on.
     #[error("Unable to allocate stream id")]
     UnableToAllocStreamId,
+
+    /// Prepared statement id changed after repreparation.
```
:camping:  Could we reorder variants of this enum so that they are grouped and sorted according to the moment in the request execution they can arise at?
I.e., serialization errors would come before deserialization errors, and `BodyExtensionsParseError` would go before `DbError` and `CqlResultParseError`.

This way it is easier to follow the possible errors along the request execution path.

wprzytula commented on scylla/src/transport/load_balancing/mod.rs:
```diff
@@ -33,31 +33,31 @@ pub struct RoutingInfo<'a> {
 /// the best to contact (e.g. they might have the lowest latency).
 ///
-/// Most queries are sent on the first try, so the query execution layer rarely needs to know more
+/// Most queries are sent on the first try, so the request execution layer rarely needs to know more
```
:wrench: Please replace the plural version, too:
`queries` -> `requests`

wprzytula commented on scylla/src/transport/load_balancing/mod.rs:
```diff
@@ -33,31 +33,31 @@ pub struct RoutingInfo<'a> {
 ///
-/// When a query is prepared to be sent to ScyllaDB/Cassandra, a `LoadBalancingPolicy`
+/// When a request is prepared to be sent to ScyllaDB/Cassandra, a `LoadBalancingPolicy`
 /// implementation constructs a load balancing plan. That plan is a list of
 /// targets (target is a node + an optional shard) to which
-/// the driver will try to send the query. The first elements of the plan are the targets which are
+/// the driver will try to send the request. The first elements of the plan are the targets which are
 /// the best to contact (e.g. they might have the lowest latency).
```
:100: Thanks for this naming clarification!

wprzytula commented on scylla/src/transport/session.rs:
```diff
@@ -2151,7 +2151,7 @@ where
                         trace!(parent: &span, "Query succeeded");
```
:wrench: `Query` -> `Request`

wprzytula commented on scylla/src/transport/session.rs:
```diff
@@ -2165,7 +2165,7 @@ where
                             "Query failed"
```
:wrench: `Query` -> `Request`

wprzytula commented on scylla/src/transport/iterator.rs:
```diff
@@ -217,7 +217,7 @@ where
                             error = %error,
                             "Query failed"
                         );
```
:wrench: :thought_balloon:  Here I'm not sure if we need the name change or not. After all, `QueryPager`, as the name suggests, should be mainly used for queries, so we can assume such usage in the naming.
OTOH Session is no longer going to log `Query failed`, but rather `Request failed`.

muzarski replied:
I decided to change it (to be consistent with session.rs).

wprzytula commented on scylla/src/transport/retry_policy.rs:
```diff
@@ -3,12 +3,12 @@
 
-/// Information about a failed query
-pub struct QueryInfo<'a> {
+/// Information about a failed request
+pub struct RequestInfo<'a> {
```
:pick: I'd appreciate this name change in a separate commit, because now it interferes with (rather small) logic changes due to `QueryError` -> `UserRequestError` change.

wprzytula replied:
(no need to extract the commit now, it's more of a note for the future)

muzarski commented:
Addressed @wprzytula comments

muzarski commented:
Rebased on main

Lorak-mmk commented on scylla/src/transport/errors.rs:
```diff
@@ -908,6 +908,8 @@ pub(crate) enum UserRequestError {
+    SerializationError(#[from] SerializationError),
 }
 
 impl UserRequestError {
```
Commit: "errors: Add serialization error to UserRequestError "

🤔 At this point I'd like comment on `QueryError` and `UserRequestError` that explain what they are for and what is the difference. I'm not 100% sure what is their intended meaning, which makes the changes a bit more difficult to understand.
My current understanding is that "UserRequestError" is an error that happened during a single try of a request execution (single try = no retries or speculative executions). Is that correct?

If it is correct, could you explain why you chose "**User**RequestError"? 
Who is the "user"?
User of the library? Probably not, because our internal queries in topology use that code too, right?
The caller? Sure, but then why "QueryError" does not have "User" in  the name?

If my understanding is correct, maybe "RequestAttemptError" would be a better name?

muzarski replied:
>  My current understanding is that "UserRequestError" is an error that happened during a single try of a request execution (single try = no retries or speculative executions). Is that correct?

Correct.

> If it is correct, could you explain why you chose "**User**RequestError"? Who is the "user"? User of the library? Probably not, because our internal queries in topology use that code too, right? The caller? Sure, but then why "QueryError" does not have "User" in the name?

User of the library. I know that this name is not perfect. Probably the best choice would be to ditch the `User` prefix. However, we would need to rename the internal `RequestError` enum which represents a failure of a single attempt of ANY CQL request (including the ones sent during connection initialization - see `ConnectionSetupRequestError`). But since `RequestError` is currently a `pub(crate)` error type (and I do not plan to change that), I think we can safely rename it (to something like `RequestErrorInternal`)  if we don't like the **User** prefix in the public error type.

> If my understanding is correct, maybe "RequestAttemptError" would be a better name?

This (adding `Attempt` to error name) is actually done in https://github.com/scylladb/scylla-rust-driver/pull/1159/commits/1aab889cc9f08522aab811e01c76910a7a39f456 of follow-up [PR](https://github.com/scylladb/scylla-rust-driver/pull/1159)

Lorak-mmk commented on scylla/src/transport/errors.rs:
```diff
@@ -871,32 +871,52 @@ pub enum CqlEventHandlingError {
+/// on this error.
 #[derive(Error, Debug)]
-pub(crate) enum UserRequestError {
+#[non_exhaustive]
```
Commit: "errors: pubify UserRequestError"

🤔 Okay, so this comment confirms my understanding - and I'm even more sure that the name should be changed.

Lorak-mmk commented on scylla/src/transport/downgrading_consistency_retry_policy.rs:
```diff
@@ -49,11 +49,11 @@ impl Default for DowngradingConsistencyRetrySession {
-                    QueryError::DbError(DbError::Unavailable { .. }, _) => {
+                    UserRequestError::DbError(DbError::Unavailable { .. }, _) => {
                         // JAVA-764: if the requested consistency level is serial, it means that the operation failed at
                         // the paxos phase of a LWT.
```
Commit: "retry_policy: decision based on UserRequestError"

🔧 There seem to be 2 changes combined in this commit:
- Narrowing the type, as the description says
- Renaming "QueryInfo" to "RequestInfo". Can those be separated?

Lorak-mmk commented on scylla/src/transport/downgrading_consistency_retry_policy.rs:
```diff
@@ -49,11 +49,11 @@ impl Default for DowngradingConsistencyRetrySession {
                 return match query_info.error {
-                    QueryError::DbError(DbError::Unavailable { .. }, _) => {
+                    UserRequestError::DbError(DbError::Unavailable { .. }, _) => {
                         // JAVA-764: if the requested consistency level is serial, it means that the operation failed at
```
🔧 You changed the type of the argument, but not the name. Could you change the name too (here and in other places)?

Lorak-mmk commented on scylla/src/transport/downgrading_consistency_retry_policy.rs:
```diff
@@ -49,10 +49,10 @@ impl Default for DowngradingConsistencyRetrySession {
-                return match query_info.error {
+                return match request_info.error {
                     UserRequestError::DbError(DbError::Unavailable { .. }, _) => {
                         // JAVA-764: if the requested consistency level is serial, it means that the operation failed at
```
Ah, you changed the name in this commit - this makes my previous commit obsolete.
However, I still think that "QueryInfo" -> "RequestInfo" rename should be moved into this commit (or into its own commit).

muzarski replied:
Moved to `retry_policy: replace query mentions with request` commit

muzarski commented:
Rebased on main (modules refactor)

muzarski commented:
Flaky `test_coalescing` failed

mykaul commented:
> Flaky `test_coalescing` failed

Is there an open issue about it? Reference it please (or open one)

muzarski commented:
> > Flaky `test_coalescing` failed
> 
> Is there an open issue about it? Reference it please (or open one)

https://github.com/scylladb/scylla-rust-driver/issues/864

muzarski commented:
Rebased on main

muzarski commented:
v1.2: Renamed `UserRequestError` -> `RequestAttemptError` (commit taken from follow-up PR)
